doc/tutorial/network_ring.rst

```diff
@@ -0,0 +1,337 @@
+      # must be implemented.
+      def num_cells(self):
+         return self.ncells
+
+      # (7) The cell_description method returns a cell
+      def cell_description(self, gid):
+         return make_cable_cell(gid)
+
+      # The kind method returns the type of cell with gid.
+      # Note: this must agree with the type returned by cell_description.
+      def cell_kind(self, gid):
+         return arbor.cell_kind.cable
+
+      # (8) Make a ring network
+      def connections_on(self, gid):
+         src = (gid-1)%self.ncells
+         w = 0.01
+         d = 5
+         return [arbor.connection(arbor.cell_member(src,0), arbor.cell_member(gid,0), w, d)]
+
+      def num_targets(self, gid):
+         return 1
+
+      def num_sources(self, gid):
+         return 1
+
+      # (9) Attach a generator to the first cell in the ring.
+      def event_generators(self, gid):
+         if gid==0:
+               sched = arbor.explicit_schedule([1])
+               return [arbor.event_generator(arbor.cell_member(0,0), 0.1, sched)]
+         return []
+
+      def probes(self, gid):
+         return [arbor.cable_probe_membrane_voltage('"root"')]
+
+      def global_properties(self, kind):
+         return self.props
+
+The execution context
+*********************
+
+The execution context contains all system-specific information needed by the simulation: it contains the
+thread pool which handles multi-threaded optimization on the CPU; it knows the relevant GPU attributes
+if a GPU is available; and it holds the MPI communicator for distributed simulations.
+The details of the execution context can be customized by the user. However,
+the ideal settings can usually be inferred from the system.
+
+In step **(10)** we specify the number of threads in the thread pool and
+let Arbor know that we have no preference for a particular GPU.
+
+Printing out the :class:`arbor.context` object will show some information of the
+capabilities of the system and the way it was configured. Most Arbor objects can
+be printed out to get obtain some information about its configuration.
+
+.. code-block:: python
+
+   # (10) Set up the hardware context
+   context = arbor.context(threads=12, gpu_id=None)
+   print(context)
+
+The meter manager
+*****************
+
+Understanding where the execution of the simulation spends time and memory is helpful for optimisation.
+The :class:`arbor.meter_manager` facilitates this. You can set one up per
+:class:`arbor.context` and set :func:`checkpoints <arbor.meter_manager.checkpoint>`
+wherever you need them.
+
+Step **(11)** creates the meters object.
+
+Step **(12)** instantiates our recipe with 4 cells. Then, we set the first checkpoint.
+
+.. code-block:: python
+
+   # (11) Set up and start the meter manager
+   meters = arbor.meter_manager()
+   meters.start(context)
+
+   # (12) Instantiate recipe
+   ncells = 4
+   recipe = ring_recipe(ncells)
+   meters.checkpoint('recipe-create', context)
+
+The domain decomposition
```

Comment:
The domain decomposition is not required in the Python interface, and I fear that it is a bit of a distraction here.

How about the following idea: don't discuss hardware contexts or domain decomposition in this example. Mention that the next example/tutorial will cover the topic of GPUs, MPI, threads and distributing work over them.

Then that example can take our simple ring model, and add these features, once the reader has comfortably set up their first recipe example?

Reply:
I can break it out, together with the manager. I actually started on the MPI version of this network, so I'll introduce it there.